pkg/groupsync/groups.go

```diff
@@ -0,0 +1,187 @@
+	Attributes any `json:"attributes,omitempty"`
+}
+
+// Group represents a group in a group system.
+type Group struct {
+	// ID is the group's ID in the group system.
+	ID string `json:"id,omitempty"`
+	// Attributes represent arbitrary attributes about the group
+	// in the given group system. This field is typically set by
+	// the corresponding ReadGroupClient when retrieving the group.
+	Attributes any `json:"attributes,omitempty"`
+}
+
+// Member represents a member of a group. A member may either be
+// a User or another Group. An instance of Member always be either
+// a User or a Group but not both.
+type Member interface {
+	// IsGroup returns whether this Member is a Group.
+	IsGroup() bool
+
+	// IsUser returns whether this Member is a User.
+	IsUser() bool
+
+	// Group returns the underlying group if this Member is a group.
+	// Otherwise, it returns an error.
+	Group() (*Group, error)
+
+	// User returns the underlying user if this Member is a user.
+	// Otherwise, it returns an error.
+	User() (*User, error)
+}
+
+// UserMember represents a user membership of a group.
+type UserMember struct {
+	Usr *User
+}
+
+// IsUser returns whether this Member is a User. Always returns true.
+func (u *UserMember) IsUser() bool {
+	return true
+}
+
+// IsGroup returns whether this Member is a Group. Always returns false.
+func (u *UserMember) IsGroup() bool {
+	return false
+}
+
+// Group returns an error.
+func (u *UserMember) Group() (*Group, error) {
+	return nil, fmt.Errorf("user is not a group")
+}
+
+// User returns the underlying user if this Member.
+func (u *UserMember) User() (*User, error) {
+	return u.Usr, nil
+}
+
+// GroupMember represents a group membership of a group.
+type GroupMember struct {
+	Grp *Group
+}
+
+// IsGroup returns whether this Member is a Group. Always returns true.
+func (g *GroupMember) IsGroup() bool {
+	return true
+}
+
+// IsUser returns whether this Member is a User. Always returns false.
+func (g *GroupMember) IsUser() bool {
+	return false
+}
+
+// Group returns the underlying group of this Member.
+func (g *GroupMember) Group() (*Group, error) {
+	return g.Grp, nil
+}
+
+// User returns an error.
+func (g *GroupMember) User() (*User, error) {
+	return nil, fmt.Errorf("group is not a user")
+}
+
+// Descendants retrieve all users (children, recursively) of the given
+// group ID using the given memberFunc. This function serves mostly as
+// a utility function when implementing ReadGroupClients for when there
+// is no special logic for fetching descendants.
+func Descendants(ctx context.Context, groupID string, memberFunc func(context.Context, string) ([]Member, error)) ([]*User, error) {
+	members, err := memberFunc(ctx, groupID)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching group members: %s, %w", groupID, err)
+	}
+	var users []*User
+	var merr error
+	for _, member := range members {
+		if member.IsUser() {
+			user, _ := member.User()
```

Comment:
Lint didn't give you problem about not checking err?
And why can we safely ignore the error here?

Reply:
We can ignore the error as the check `member.IsUser()` was true in order to enter the `if` block. so we will never get an error as the member will always be a user at that point.

Reply:
And yeah the linter didn't complain about ignoring the err. It might be smart enough to realize the error will never occur.